### PR TITLE
Add memorization of tool-shared dynamics options (fix #4007)

### LINKED ISSUE #4007: Dynamics option fixes
- [ ] **Important** Don't save/set the properties of all tools when the "Save in all Tools" is checked, we could use a shared tool section (`[tool.shared.dynamics]`?) for that. Because unchecking the "same in all tools" should not keep the dynamics on all tools.
- [ ] Use the matrix name/ID (instead of combobox index) to save the selected matrix for dynamic dithering
- [ ] Reset dynamics option from `Preferences::resetToolPreferences()` (there is [a comment about this in pref.xml](https://github.com/aseprite/aseprite/blob/572cdf2b0ee9deb4a21acea72ddcf47a8601e94e/data/pref.xml#L448))

### PR DESCRIPTION
**Pull request by Gasparoken**
fix #4007

### CODE REVIEW

**Comment by Gasparoken** [comment]
First approach to solve first point of #4007.
I'll try to implement this https://github.com/aseprite/aseprite/issues/4007#issuecomment-1684317298

**Comment by dacap** [comment]
I think there is no need to duplicate the shared preferences, probably we could use the `Preferences::tool(nullptr)` to reference the shared options.

**Comment by Gasparoken** [comment]
> I think there is no need to duplicate the shared preferences, probably we could use the `Preferences::tool(nullptr)` to reference the shared options.

I am working on it, today I'll push separate commits for each point of #4007.

**Comment by Gasparoken** [comment]
Ready for review.

File: src/app/pref/preferences.cpp [hunk 1]
```diff
@@ -139,7 +139,20 @@ bool Preferences::isSet(OptionBase& opt) const
...
+      return *it->second;
+    }
+    else {
+      std::string section = std::string("tool.shared");
+      ToolPreferences* toolPref = new ToolPreferences(section);
+      m_tools["shared"] = toolPref;
+      toolPref->load();
+      return *toolPref;
+    }
+  }
```
Comment (dacap) [inline src/app/pref/preferences.cpp:155]: As this code is similar to the one that follows, probably we can create an `id` local variable and do something like this:

```suggestion
 std::string id;
 
  // If tool == nullptr it means that the tool's options are shared with all tools.
  if (tool)
    id = tool->getId();
  else
    id = "shared";
```

and use `id` instead of `tool->getId()` in the following code, and check for `tool` before using `tool->`.

File: src/app/ui/dithering_selector.cpp [hunk 2]
```diff
@@ -198,7 +198,14 @@ void DitheringSelector::onInitTheme(ui::InitThemeEvent& ev)
     setSizeHint(getItem(0)->sizeHint());
 }
 
-void DitheringSelector::regenerate()
+void DitheringSelector::setSelectedItemByName(std::string name)
```
Comment (dacap) [inline src/app/ui/dithering_selector.cpp:201]:
```suggestion
void DitheringSelector::setSelectedItemByName(const std::string& name)
```

File: src/app/ui/dithering_selector.cpp [hunk 3]
```diff
@@ -255,9 +262,9 @@ void DitheringSelector::regenerate()
       }
       break;
   }
-
-  setSelectedItemIndex(0);
-  setSizeHint(getItem(0)->sizeHint());
+  selectedItemIndex = std::min(std::max(0, selectedItemIndex), getItemCount()-1);
```
Comment (dacap) [inline src/app/ui/dithering_selector.cpp:265]:
```suggestion
  selectedItemIndex = base::clamp(selectedItemIndex, 0, std::max(0, getItemCount()-1));
```
We need to include "base/clamp.h" for this.

File: src/app/ui/dithering_selector.h [hunk 4]
```diff
@@ -27,13 +27,13 @@ namespace app {
 
     render::DitheringAlgorithm ditheringAlgorithm();
     render::DitheringMatrix ditheringMatrix();
+    void setSelectedItemByName(std::string name);
```
Comment (dacap) [inline src/app/ui/dithering_selector.h:30]:
```suggestion
    void setSelectedItemByName(const std::string& name);
```

File: src/app/ui/dynamics_popup.cpp [hunk 5]
```diff
@@ -289,15 +306,20 @@ void DynamicsPopup::setOptionsGridVisibility(bool state)
...
 }
 
-int DynamicsPopup::ditheringIndex() const {
+std::string DynamicsPopup::ditheringMatrixName() const {
   if (m_ditheringSel)
-    return m_ditheringSel->getSelectedItemIndex();
-  return 0;
+    return m_ditheringSel->getItemText(
+      m_ditheringSel->getSelectedItemIndex());
+  return Strings::dithering_selector_no_dithering();
```
Comment (dacap) [inline src/app/ui/dynamics_popup.cpp:313]: Not sure if this should return an empty string in this case:
```suggestion
  return std::string();
```

File: src/ui/combobox.h [hunk 6]
```diff
@@ -75,6 +75,7 @@ namespace ui {
 
     int getSelectedItemIndex() const;
     void setSelectedItemIndex(int itemIndex);
+    int getSelectedIndexByName(std::string itemName) const;
```
Comment (dacap) [inline src/ui/combobox.h:78]:
```suggestion
    int getSelectedIndexByName(const std::string& itemName) const;
```

File: src/app/ui/dynamics_popup.cpp [hunk 7]
```diff
@@ -50,6 +52,8 @@ enum {
 
 } // anonymous namespace
 
+static bool g_same_toggle = false;
+
```
Comment (dacap) [inline src/app/ui/dynamics_popup.cpp:56]: Unused variable:
```suggestion
```

**Review by dacap — CHANGES_REQUESTED** [review]

File: src/app/ui/dithering_selector.cpp (hunk 3 above)
Comment (Gasparoken) [reply src/app/ui/dithering_selector.cpp:265]: I used `std::clamp` instead.

Comment (dacap) [reply src/app/ui/dithering_selector.cpp:265]: Oh yes, now we can use it 👍

**Comment by Gasparoken** [comment]
Squashed, and ready for review.

**Review by dacap — CHANGES_REQUESTED** [review]

File: src/ui/combobox.cpp [hunk 8]
```diff
@@ -312,6 +312,14 @@ void ComboBox::setSelectedItemIndex(int itemIndex)
   }
 }
 
+int ComboBox::getSelectedIndexByName(const std::string& itemName) const
```
[the remark this answers is not in the document]
Comment (Gasparoken) [reply src/ui/combobox.cpp:315]: Yes! You're right, I'll use `findItemIndex()` instead.

File: data/pref.xml [hunk 9]
```diff
@@ -317,9 +317,9 @@
       <option id="discard_brush" type="bool" default="false" />
     </section>
     <section id="shared">
-      <option id="share_dynamics" type="bool" default="false" />
       <option id="share_ink" type="bool" default="false" />
       <option id="ink" type="app::tools::InkType" default="app::tools::InkType::DEFAULT" />
+      <option id="share_dynamics" type="bool" default="true" />
```
Comment (dacap) [inline data/pref.xml:322]: Not sure why this line changed to this place now, but you can just change `false` to `true`.

File: src/app/ui/dynamics_popup.cpp [hunk 10]
```diff
@@ -289,15 +304,20 @@ void DynamicsPopup::setOptionsGridVisibility(bool state)
     expandWindow(sizeHint());
 }
 
-int DynamicsPopup::ditheringIndex() const {
+std::string DynamicsPopup::ditheringMatrixName() const {
```
Comment (dacap) [inline src/app/ui/dynamics_popup.cpp:307]:
```suggestion
std::string DynamicsPopup::ditheringMatrixName() const
{
```

File: src/app/ui/dynamics_popup.cpp [hunk 11]
```diff
@@ -289,15 +304,20 @@ void DynamicsPopup::setOptionsGridVisibility(bool state)
...
-  return 0;
+    return m_ditheringSel->getItemText(
+      m_ditheringSel->getSelectedItemIndex());
+  return std::string();
 }
 
-void DynamicsPopup::loadDynamicsPref() {
-  auto& dynaPref = Preferences::instance().tool(
-    App::instance()->activeTool()).dynamics;
+void DynamicsPopup::loadDynamicsPref(bool sameInAllTools) {
```
Comment (dacap) [inline src/app/ui/dynamics_popup.cpp:314]:
```suggestion
void DynamicsPopup::loadDynamicsPref(bool sameInAllTools)
{
```

File: src/app/ui/dynamics_popup.cpp [hunk 12]
```diff
@@ -324,10 +344,33 @@ void DynamicsPopup::loadDynamicsPref() {
            dynaPref.gradient() == tools::DynamicSensor::Velocity);
 
   if (m_ditheringSel)
-    m_ditheringSel->setSelectedItemIndex(dynaPref.matrixIndex());
+    m_ditheringSel->setSelectedItemByName(dynaPref.matrixName());
+}
+
+void DynamicsPopup::saveDynamicsPref(bool sameInAllTools) {
```
Comment (dacap) [inline src/app/ui/dynamics_popup.cpp:350]:
```suggestion
void DynamicsPopup::saveDynamicsPref(bool sameInAllTools)
{
```

**Review by dacap — CHANGES_REQUESTED** [review]
Just some minor changes, and remember to include fix #number of issue in the commit summary itself.

**Comment by Gasparoken** [comment]
Ready for merge.

**Comment by dacap** [comment]
Testing this, it works in a strange way (I think checking/unchecking "Same in all tools" option shouldn't modify the visible preferences of the active tool):

https://github.com/aseprite/aseprite/assets/39654/1ed563fe-703f-4e4d-82b5-6d4d1279d40c

The "Same in all tools" checkbox is switching between all the options between the active tool and the shared values. I think the checkbox shouldn't change the values.

This is 1) if I set some dynamics and the "Same in all tools" is unchecked, checking it should copy the active config as the shared values, and 2) If "Same in all tools" is checked and I uncheck it, it should copy the shared values to the active tool config (without touching other tools), in this way the visible options of the active tool will not be modified.

What do you think about this?

**Comment by Gasparoken** [comment]
> This is 1) if I set some dynamics and the "Same in all tools" is unchecked, checking it should copy the active config as the shared values, and 2) If "Same in all tools" is checked and I uncheck it, it should copy the shared values to the active tool config (without touching other tools), in this way the visible options of the active tool will not be modified.
> 
> What do you think about this?

The current behavior is like a switch between the particular tool settings and shared settings as a separate memories.
I did it in that way to not destroy the settings of the tool: 
- This allows to use 2 configs as [shared set] <-> [tool's specific set] toggle button.
- This avoids to destroy the configuration of other tools accidentally by pressing "Same in all tools" + closing the Dynamics window.
- The drawback as you said: this can be a strange behavior for artists.

Maybe a different UI can improve the idea:
Sprite-0002


If this behavior is strange, so will the following case:
This implementation doesn't consider to save the `maxSize` and `maxAngle` as a `DynamicsOptions` since both settings are saved always as per-tool setting. So, the `DynamicsOptions` with "Same in all tools" marked won't have the same settings in all tools because each tool has its own `maxSize` and `maxAngle` setting.
Again, this is in this way to prioritize settings preservation, but I understand it could be confusing.
@dacap I'll need your opinion if I have to change the implementation to do some special with `maxSize` and `maxAngle`.

In conclusion of all this I think:
1. Should Brush Size/Brush Angle/Brush Type have a "Same in all tools" check box?
2. Implement something better in all ways: save the settings just as the Custom Brushes.
3. For the current commit, I'll modify the "Same in all tools" behavior like you said.

**Comment by dacap** [comment]
> Should Brush Size/Brush Angle/Brush Type have a "Same in all tools" check box?

No right now (there is a plan to redesign the brush popup completely).

> Implement something better in all ways: save the settings just as the Custom Brushes.

No, this is too much for the moment and I think we have to still find a right design/UI for all these problems (redesigning the brush popup). It's something for the future.

> For the current commit, I'll modify the "Same in all tools" behavior like you said.

👍 

As a side note: I know that you squashed the commits in the past, but it would be great to reduce the commits in a future (e.g. the matrix name change could be a first PR/commit so it's not anymore in this PR and already merged). Right now you can continue all the work in this one, but for a future making smaller PRs/commits for each change will be a benefit in case that a commit must re-reviewed.

**Comment by Gasparoken** [comment]
Ready.
